job-runner/chain/reaper.go

```diff
@@ -0,0 +1,642 @@
+
+	// Copy job data from this job to all child jobs. When a job has multiple
+	// parent jobs, it will get jobData copied from all of its parent jobs in
+	// the order that they finish. Therefore, if multiple parent jobs try to set
+	// the same jobData field, it is undetermined which parent job the job will
+	// receive the final copy of that jobData field from.
+	for _, nextJob := range r.chain.NextJobs(job.Id) {
+		nextJLogger := jLogger.WithFields(log.Fields{"next_job_id": nextJob.Id})
+
+		for k, v := range job.Data {
+			nextJob.Data[k] = v
+		}
+		r.chainRepo.Set(r.chain)
+
+		// Enqueue any runnable child jobs.
+		if !r.chain.IsRunnable(nextJob.Id) {
+			nextJLogger.Infof("next job is not runnable - not enqueuing it")
+			continue
+		}
+		nextJLogger.Infof("next job is runnable - enqueuing it")
+		if r.chain.IsSequenceStartJob(nextJob.Id) {
+			// Starting a sequence, so increment sequence try count.
+			r.chain.IncrementSequenceTries(nextJob.Id)
+			seqLogger := r.logger.WithFields(log.Fields{"sequence_id": job.SequenceId, "sequence_try": r.chain.SequenceTries(job.Id)})
+			seqLogger.Info("starting try of sequence")
+		}
+		r.runJobChan <- nextJob
+	}
+}
+
+// Finalize determines the final state of the chain and sends it to the Request
+// Manager.
+func (r *RunningChainReaper) Finalize(complete bool) {
+	if complete {
+		r.logger.Infof("chain is done, all jobs finished successfully")
+		r.chain.SetState(proto.STATE_COMPLETE)
+	} else {
+		r.logger.Warn("chain is done, some jobs failed")
+		r.chain.SetState(proto.STATE_FAIL)
+	}
+	r.chainRepo.Set(r.chain)
+	if r.sendFinalState() {
+		r.chainRepo.Remove(r.chain.RequestId())
+	}
+}
+
+// -------------------------------------------------------------------------- //
+
+// Job Reaper for chains that are being suspended (stopped to be resumed later).
+type SuspendedChainReaper struct {
+	reaper
+	runnerRepo runner.Repo
+}
+
+// Run reaps jobs when they finish running. For each job reaped, if it's...
+// - completed: prepare subsequent jobs (copy jobData).
+// - failed:    prepare a sequence retry.
+// - stopped:   do nothing (job will be retried when chain is resumed).
+func (r *SuspendedChainReaper) Run() {
+	defer close(r.doneChan)
+
+	// If there are already no jobs left to reap, the running reaper must have
+	// finished and finalized the chain before it got switched out for this reaper.
+	// There's nothing left to do, so return right away.
+	if r.done() {
+		return
+	}
+
+REAPER:
+	for {
+		select {
+		case job := <-r.doneJobChan:
+			r.Reap(job)
+
+			// If runnerRepo is consistently empty, there are no more running jobs
+			// and thus no more jobs to be reaped. Break out of the reaper loop.
+			if r.runnerRepo.Count() == 0 {
+				// runnerRepo might be empty very briefly after a job is sent to
+				// runJobChan but before its runner is created, so wait a while
+				// and double check.
+				time.Sleep(500 + time.Millisecond)
+				if r.runnerRepo.Count() == 0 {
```

Comment:
This logic is duplicated in `done()` so let that be single source of truth. Then here:
```
if r.done() {
    break REAPER
}
```